### PR TITLE
refactor: improve underlying goroutine termination of timeout reader

### PR DESCRIPTION
**Pull request by mrekucci**
### Checklist

- [x] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [ ] My change requires a documentation update, and I have done it.
- [ ] I have added tests to cover my changes.
- [x] I have filled out the description and linked the related issues.

### Description
Improve timeout reader:
- don't terminate the underlying goroutine when zero bytes are read
- terminate underlying goroutine on any error encounter

### CODE REVIEW

**Comment by mrekucci** [comment]
> * goroutine in `TimeoutReader` function should be updated accordingly

Can you be more specific?

> * switch statement in `Read` seems complex, why have you decided to go along that way, compared to previous approach of just writing data back to timeout goroutine?

The `Stop` call has been moved to the main goroutine to avoid possible data-race. According to the [documentation](https://pkg.go.dev/time#Timer.Stop), it is also recommended to empty the timer after calling the `Stop` method.

File: pkg/util/ioutil/ioutil.go [hunk 1]
```diff
@@ -25,7 +24,7 @@ func (tr *timeoutReader) Read(p []byte) (int, error) {
 	n, err := tr.r.Read(p)
 
 	switch {
-	case (n == 0 || errors.Is(err, io.EOF)) && tr.t.Stop():
+	case err != nil && tr.t.Stop():
```
Comment (istae) [inline pkg/util/ioutil/ioutil.go:27]: can we write this function without the `fallthrough`

**Comment by mrekucci** [comment]
> > Can you be more specific?
> 
> goroutine created in `TimeoutReader` should also be updated with same fix, namely, switch and cases should be removed.

Not sure I follow, can you provide some code?

> > The `Stop` call has been moved to the main goroutine to stop the timer as soon as possible and to avoid possible data-race. According to the [documentation](https://pkg.go.dev/time#Timer.Stop), it is also recommended to empty the timer after calling the `Stop` method.
> 
> Have a look at this PR, it seem to have simpler fix for all problems in this utility #3894

PR #3894 looks overly complicated, without context the timeout reader can be simplified and just simply leverage `time.AfterFunc`.

File: pkg/util/ioutil/ioutil.go (hunk 1 above)
Comment (mrekucci) [reply pkg/util/ioutil/ioutil.go:27]: Done.

**Review by istae — APPROVED** [review]

**Review by vladopajic — APPROVED** [review]

**Pull request merged**